torch/csrc/autograd/engine.cpp

```diff
@@ -747,34 +729,58 @@ variable_list Engine::execute_with_graph_task(
 
   // Not a worker
   if (worker_device == NO_DEVICE) {
-    // Wait for all tasks to complete
-    graph_task->not_done_.wait(
-        lock, [&graph_task] { return graph_task_completed(*graph_task); });
+    // graph_task_exec_post_processing is done when the Future is marked as
+    // completed in mark_graph_task_completed.
+    return graph_task->future_result_;
   } else {
     graph_task->owner_ = worker_device;
-    ++total_depth;
-    if(current_depth >= max_recursion_depth_){
+    if (current_depth >= max_recursion_depth_) {
       // See Note [Reentrant backwards]
       // If reached the max depth, switch to a different thread
       add_thread_pool_task(graph_task);
-      graph_task->not_done_.wait(
-          lock, [&graph_task] { return graph_task_completed(*graph_task); });
+      // graph_task_exec_post_processing is done when the Future is marked as
+      // completed in mark_graph_task_completed.
+      return graph_task->future_result_;
     } else {
+      // Total depth needs to be updated only in this codepath, since it is
+      // not used in the block above (when we call add_thread_pool_task).
```

Comment:
nit: maybe mention that `graphtask.reentrant_depth_` is used instead of the TLS as we will change thread in the other branch.